src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1993,6 +2160,52 @@ def translate_device(dev):
         else:
             return False
 
+    @private
+    def get_iommu_type(self):
+        def check_iommu(iommu_type):
+            IOMMU_TEST = {'VT-d': {'arglist': ['/usr/sbin/acpidump', '-t'],
+                                   'string': 'DMAR'},
+                          'amdvi': {'arglist': ['/sbin/sysctl', 'hw'],
+                                    'string': 'vmm.amdvi.enable: 1'}}
+            if iommu_type in IOMMU_TEST:
+                proc1 = None
+                proc2 = None
+                try:
+                    proc1 = subprocess.Popen(IOMMU_TEST[iommu_type]['arglist'],
+                                             stdout=subprocess.PIPE,
+                                             stderr=subprocess.DEVNULL)
+                    proc2 = subprocess.Popen(['/usr/bin/grep', IOMMU_TEST[iommu_type]['string']],
+                                             stdin=proc1.stdout,
+                                             stdout=subprocess.PIPE,
+                                             stderr=subprocess.DEVNULL)
+                    proc1.stdout.close()  # Allow proc1 to receive a SIGPIPE if proc2 exits.
+                    outs, errs = proc2.communicate(timeout=5)
+                except (subprocess.TimeoutExpired, OSError, ValueError) as e:
+                    if proc2 is not None:
+                        proc2.kill()
+                        outs, errs = proc2.communicate()
+                    elif proc1 is not None:
+                        proc1.kill()
+                    self.logger.warn(f'An error occured when checking for iommu ({iommu_type}). {e}.')
+                else:
+                    if proc1.poll() is None:  # make sure proc1 has terminated
+                        proc1.kill()
+                    if proc1.returncode < 1 and proc2.returncode < 2:
```

Comment:
Sam here about returncode, can you please confirm if all return codes below these are valid ?

Reply:
See above. For grep (proc2) return codes 0 and 1 are good. Done.